clouddriver-sql/src/main/kotlin/com/netflix/spinnaker/clouddriver/sql/event/SqlEventRepository.kt

```diff
@@ -0,0 +1,298 @@
+   *
+   * This is a bit wonky: In the case of [ComposedSpinnakerEvent]s, we want to initialize the event so we can
+   * correctly serialize it, but we don't want to increment the sequence for these events as they aren't
+   * actually on the event log yet. If we're in a [ComposedSpinnakerEvent], we just provide a "-1" sequence
+   * number and a real, valid sequence will be assigned if/when it gets saved to the event log.
+   */
+  private fun SpinnakerEvent.initialize(
+    aggregateType: String,
+    aggregateId: String,
+    originatingVersion: Long,
+    currentSequence: Long?
+  ): Long? {
+    var nextSequence = if (currentSequence != null) {
+      currentSequence + 1
+    } else {
+      null
+    }
+
+    // timestamp is calculated on the SQL server
+    setMetadata(EventMetadata(
+      id = UUID.randomUUID().toString(),
+      aggregateType = aggregateType,
+      aggregateId = aggregateId,
+      sequence = nextSequence ?: -1,
+      originatingVersion = originatingVersion,
+      serviceVersion = serviceVersion.resolve()
+    ))
+
+    if (this is CompositeSpinnakerEvent) {
+      this.getComposedEvents().forEach { event ->
+        // We initialize composed events with a null sequence, since they won't actually get added to the log at
+        // this point; that's up to the action to either add it or not, at which point it'll get a sequence number
+        event.initialize(aggregateType, aggregateId, originatingVersion, null)?.let {
+          nextSequence = it
+        }
+      }
+    }
+
+    return nextSequence
+  }
+
+  override fun list(aggregateType: String, aggregateId: String): List<SpinnakerEvent> {
+    return withPool(POOL_NAME) {
+      jooq.select().from(EVENTS_TABLE)
+        .where(field("aggregate_type").eq(aggregateType)
+          .and(field("aggregate_id").eq(aggregateId)))
+        .orderBy(field("sequence").asc())
+        .fetchEvents(objectMapper)
+    }
+  }
+
+  override fun listAggregates(criteria: ListAggregatesCriteria): EventRepository.ListAggregatesResult {
+    // TODO(rz): validate criteria
+
+    return withPool(POOL_NAME) {
+      val conditions = mutableListOf<Condition>()
+      criteria.aggregateType?.let { conditions.add(field("aggregate_type").eq(it)) }
+      criteria.token?.let { conditions.add(field("token").greaterThan(it)) }
+
+      val perPage = criteria.perPage.coerceAtMost(10_000)
+
+      val aggregates = jooq.select().from(AGGREGATES_TABLE)
+        .withConditions(conditions)
+        .orderBy(field("token").asc())
+        .limit(perPage)
+        .fetchAggregates()
+
+      val remaining = jooq.selectCount().from(AGGREGATES_TABLE)
+        .withConditions(conditions)
+        .fetchOne(0, Int::class.java) - perPage
+
+      EventRepository.ListAggregatesResult(
+        aggregates = aggregates.map { it.model },
+        nextPageToken = if (remaining > 0) aggregates.lastOrNull()?.token else null
```

Comment:
Token is effectively used as an aggregate row creation timestamp, but since aggregate rows are updated in place over time without the token incrementing, is this the right thing to paginate by? (Maybe, seems similar to sorting by `buildTime` in orca, but I could also see sorting by last updated being desirable.)

Reply:
`listAggregates` isn't going to be consumed by anyone but operators / developers as far as I can tell so far. I wanted pagination because it was The Right Thing to do, and also something easy to reason about. If the pagination token were based off of something like a `lastUpdated` field, the pagination results would be inconsistent and would probably lead to some odd behavior for consumers.